### PR TITLE
Control for getty start process

### PR DESCRIPTION
**Pull request by giggsoff**
In order to not start getty process by default (from init container) we can use novel explicit getty option
in kernel command line. Which may be set using grub.cfg on the CONFIG partition as described [here](https://github.com/lf-edge/eve/blob/dfa564681838f7cd289d8f1ef00ada47e5a44ad0/docs/BOOTING.md#booting-under-uefi-firmware), i.e. by defining there `set_global dom0_extra_args "$dom0_extra_args getty "` or with just
`set_getty`.

This PR also contains `eve config mount <mountpoint>` command to support modifications of CONFIG partition content.

And several grub menu options to enable getty and clean dom0_extra_args without touching grub.cfg for one boot.

Also I implemented `debug.enable.console` option to control getty start from pillar. For now I only implemented enabling of getty,
to disable edge-node must be rebooted.

### CODE REVIEW

**Comment by mikem-zed** [comment]
I just thought maybe we should abstract GRUB options and make it easier to modify them for end user e.g. having a json.yaml file with user friendly option names ? just an idea

**Comment by giggsoff** [comment]
> I just thought maybe we should abstract GRUB options and make it easier to modify them for end user e.g. having a json.yaml file with user friendly option names ? just an idea

You mean some kind of tool to generate grub.cfg file? It might make sense.
BTW, I modified PR/commit description as we can just put `set_no_getty` into grub.cfg to call the inner function.

File: pkg/dom0-ztools/rootfs/bin/eve [hunk 1]
```diff
@@ -163,6 +182,13 @@ __EOT__
                      ;;
           esac
           ;;
+ config) case "$2" in
+               mount) mount_partlabel "CONFIG" "$3"
+                     ;;
```
Comment (eriknordmark) [inline pkg/dom0-ztools/rootfs/bin/eve:187]: If we're adding a mount it would make sense to also add an unmount.

Comment (giggsoff) [reply pkg/dom0-ztools/rootfs/bin/eve:199]: Good point, added.

**Review by eriknordmark — COMMENTED** [review]
Today we depend on having keyboard, VGA, and console available before the device is onboarded, and then having zedagent+domainmgr disable the keyboard and VGA once the device is onboarded.

Can we do that somehow and not depend on this being set at boot time?

**Comment by giggsoff** [comment]
> Today we depend on having keyboard, VGA, and console available before the device is onboarded, and then having zedagent+domainmgr disable the keyboard and VGA once the device is onboarded.
> 
> Can we do that somehow and not depend on this being set at boot time?

Is my understanding correct, that you propose to run getty on boot, but disable it later in internal logic (after some internal checks) on every boot? Or we should disable getty after onboard and to not start after subsequent reboot at all?

**Comment by eriknordmark** [comment]
> Is my understanding correct, that you propose to run getty on boot, but disable it later in internal logic (after some internal checks) on every boot? Or we should disable getty after onboard and to not start after subsequent reboot at all?

Disabling it after onboard would be desirable. 
But don't know if we want a mechanism to be able to re-enable it. We can re-enable the USB keyboard and/or VGA with the debug.enable.XXX configItems after the device has been onboarded. If we could have a debug.enable.console or debug.enable.getty following the logic we have for debug.enable.{usb,vga} that would be the best.

**Comment by eriknordmark** [comment]
> This assumes it is necessary for this to be runtime and not buildtime? For buildtime, we can remove the getty services from the yml file.

@deitch if/when we start producing separate debug and production EVE images, then it might make sense to disable it in the build for the production images. But with a single image for all purposes we can't do that.

> Finally, what is the security impact of allowing getty at all? Doesn't it mean someone can change the cmdline and have access?

In general an attacker can boot any way they want using the UEFI shell etc. But attacker can't access /persist/vault if the boot chain/arguments is different.
And other things such as /persist/* can be modified by the attacker by removing the drive to modify the content and then re-inserting it.

**Comment by giggsoff** [comment]
The main problem with making start of getty/login manageable actually comes from the reasons we use it. We want to be able to interact with console on early steps of booting, before starting of services as their start may also fail for some reason. And the way to control in such place (init section) we have is to use kernel options. We have a way to control kernel options using /config/grub.cfg. And the logic for this (option in grub.cfg) implemented in the PR.
I still thinking about the solution without touching of grub.cfg. One of the possible ways will be to delay the start of getty/login for some (e.g. minute) amount of time and try to cancel/force the start according to the logic from manageable space (pillar). Is it sounds reasonable?

**Comment by deitch** [comment]
To be clear, this change is not, "every time you want to disable/enable getty, you need to change `grub.cfg`," but rather, "this PR changes `grub.cfg` permanently, and the user on boot can select a menu option." Correct? I don't think that is bad.

**Comment by giggsoff** [comment]
> To be clear, this change is not, "every time you want to disable/enable getty, you need to change `grub.cfg`," but rather, "this PR changes `grub.cfg` permanently, and the user on boot can select a menu option." Correct? I don't think that is bad.

Well, both are not correct at least for the current code in PR (but I may change it). I did not touch any defaults. I give the way to disable getty permanently (if you will put `set_no_getty` in `grub.cfg`, also PR contains required `eve config mount` command), the way to disable for one boot (if you will select `disable getty` in grub menu) and the way to unset all options in dom0_extra_args (including permanently added  `set_no_getty` for example) using `unset dom0_extra_args` in grub menu for one boot. Tried to foresee all options.

**Comment by deitch** [comment]
Definitely confused now.

We currently don't have the ability to generate debug and prod builds (although any user can modify the source here and build, but that is out of scope).

So when you say, "modify grub.cfg", do you mean a user needs to edit it by hand when at the console?

**Comment by giggsoff** [comment]
> So when you say, "modify grub.cfg", do you mean a user needs to edit it by hand when at the console?

It is one of possibility, another is to install EVE-OS with modified grub.cfg, which in turn possible using the [logic](https://github.com/lf-edge/eve/blob/85ba1cc0d759f635ad005e1d1ff89bb0ad79a820/pkg/eve/runme.sh#L62-L67) of eve image.

**Comment by giggsoff** [comment]
I modified the logic here. 
I made init-placed getty grub-option aware (in order to have access driving by grub on edge-node which is broken in terms of pillar's start).
I implemented logic to start getty from pillar. We will start getty if node not onboarded, or if we enable `debug.enable.console` in config-properties. I have no good ideas how to gracefully stop getty/login without bunch of odd kill commands (any ideas appreciated), so disable will work only after reboot for now.

File: docs/BOOTING.md [hunk 2]
```diff
@@ -78,8 +78,11 @@ conditional or looping statements).
 Second stage GRUB is expected to do all the [heavy lifting](../pkg/grub/rootfs.cfg) of actually booting
 an EVE instance and because it resides in IMGA or IMGB partitions it can easily be upgraded and patched.
 Behavior of this stage of boot process is controlled by a read-only grub.cfg under {IMGA,IMGB}/EFI/BOOT/
-but it can further be tweaked by the grub.cfg overrides on the CONFIG partition. Note that an override
-grub.cfg is expected to be a [complete override](../pkg/grub/rootfs.cfg#L143) and anyone constructing
+but it can further be tweaked by the grub.cfg overrides on the CONFIG partition. The content of
+CONFIG partition mounted using tmpfs mount. In order to mount partition itself you can use
+`eve mount config <mountpoint>` command. Modification of files in CONFIG partition may affect booting process
```
Comment (eriknordmark) [inline docs/BOOTING.md:83]: Makes sense to mention unmount config here as well.

File: pkg/dom0-ztools/rootfs/etc/init.d/001-getty [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+# we override content of https://github.com/linuxkit/linuxkit/blob/master/pkg/getty/etc/init.d/001-getty
+# with conditional behavior
+# if we see getty in kernel command line
+# we will start getty
+
+if cat /proc/cmdline | grep -q getty; then
+  export INITGETTY=true
+  export INSECURE=true
+  /usr/bin/rungetty.sh
```
Comment (eriknordmark) [inline pkg/dom0-ztools/rootfs/etc/init.d/001-getty:11]: With this, do we still need/use linuxkit/getty:v0.5?

File: pkg/grub/rootfs.cfg [hunk 4]
```diff
@@ -308,7 +302,10 @@ function set_kvm_boot {
    set_global load_hv_cmd echo
    set_global load_dom0_cmd linux
    set_global load_initrd_cmd initrd
-   set_global dom0_extra_args "$dom0_extra_args pcie_acs_override=downstream,multifunction"
+   set_global dom0_flavor_tweaks "pcie_acs_override=downstream,multifunction"
+   if [ "$arch" = "x86_64" ]; then
+      set_global dom0_flavor_tweaks "$dom0_flavor_tweaks crashkernel=2G-64G:128M,64G-1T:256M,1T-:512M"
+   fi
```
Comment (eriknordmark) [inline pkg/grub/rootfs.cfg:308]: Currently we also set crashkernel with xen, but here you only set it with kvm. Is that what we want?

File: pkg/dom0-ztools/rootfs/etc/init.d/001-getty (hunk 3 above)
Comment (giggsoff) [reply pkg/dom0-ztools/rootfs/etc/init.d/001-getty:11]: The main idea here is to keep getty in init section to be able to debug. I put an option `enable getty` in grub menu to enable it for debugging.
Also we still rely on files and logic available in `linuxkit/getty:v0.5` from pillars part of code (I put the comment into `rootfs.yml.in` about it)

File: docs/BOOTING.md (hunk 2 above)
Comment (giggsoff) [reply docs/BOOTING.md]: Correct, modified

File: pkg/pillar/cmd/nodeagent/nodeagent.go [hunk 5]
```diff
@@ -406,6 +408,13 @@ func handleGlobalConfigImpl(ctxArg interface{}, key string,
 		ctxPtr.globalConfig = gcp
 		ctxPtr.GCInitialized = true
 	}
+	if gcp.GlobalValueBool(types.ConsoleAccess) {
+		// FIXME: explore the way to stop getty/login
+		if !ctxPtr.gettyStarted {
+			startGetty(log)
+			ctxPtr.gettyStarted = true
+		}
+	}
```
[the remark this answers is not in the document]
Comment (giggsoff) [reply pkg/pillar/cmd/nodeagent/nodeagent.go]: I moved the logic into domainmgr. Actually vga and usb depends on the logic in domainmgr, probably console option is not directly connected, but nice to keep the logic in the same place.
About your second question. Console will be available before onboarding, the same way we do it for usb/vga. And for now we need reboot to disable it (automatically or with explicit config come from controller), I put the comment about this in CONFIG-PROPERTIES.md.

**Review by eriknordmark — APPROVED** [review]
LGTM

File: pkg/grub/rootfs.cfg (hunk 4 above)
Comment (eriknordmark) [reply pkg/grub/rootfs.cfg:308]: Ah - I missed that. @rouming do you think we should keep the separate function set_crashkernel()?

**Review by eriknordmark — APPROVED** [review]
Run eden

**Comment by giggsoff** [comment]
Moved from `base.Exec().CombinedOutput()` to `exec.Command` to not hit internal timeout.

**Review by eriknordmark — APPROVED** [review]
Run eden

**Pull request merged**